### PR TITLE
SDI-4: 🐛 Correct visit / visitor mappings

### PR DESCRIPTION
**Pull request by petergphillips**

### CODE REVIEW

File: src/main/java/uk/gov/justice/hmpps/prison/api/model/VisitDetails.java [hunk 1]
```diff
@@ -49,7 +49,6 @@
     private String leadVisitor;
 
     @ApiModelProperty(value = "Relationship of main visitor to offender")
-    @JsonProperty("relationship")
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/prison/api/model/VisitDetails.java:52]: unnecessary property as no uppercase in name

File: src/main/java/uk/gov/justice/hmpps/prison/api/model/VisitDetails.java [hunk 2]
```diff
@@ -78,6 +76,10 @@
     @JsonProperty("eventOutcomeDescription")
     private String eventOutcomeDescription;
 
+    @ApiModelProperty(value = "Whether the visit was attended. Translation of eventOutcome into boolean. Defaults in NOMIS to true when the visit is created")
+    @NotBlank
+    private boolean attended;
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/prison/api/model/VisitDetails.java:81]: helper to complement `attended` on Visitor

File: src/main/java/uk/gov/justice/hmpps/prison/api/resource/BookingResource.java [hunk 3]
```diff
@@ -879,7 +879,7 @@ public AdjudicationSummary getAdjudicationSummary(@PathVariable("bookingId") @Ap
         @RequestParam(value = "fromDate", required = false) @org.springframework.format.annotation.DateTimeFormat(iso = DateTimeFormat.ISO.DATE) @ApiParam("Returned visits must be scheduled on or after this date (in YYYY-MM-DD format).") final LocalDate fromDate,
         @RequestParam(value = "toDate", required = false) @org.springframework.format.annotation.DateTimeFormat(iso = DateTimeFormat.ISO.DATE) @ApiParam("Returned visits must be scheduled on or before this date (in YYYY-MM-DD format).") final LocalDate toDate,
         @RequestParam(value = "visitType", required = false) @ApiParam(value = "Type of visit", allowableValues = "SCON, OFFI") final String visitType,
-        @RequestParam(value = "visitStatus", required = false) @ApiParam(name = "Status of visit. code from VIS_STS domain, e.g: Cancelled (CANC) or Scheduled (SCH)", example = "SCH") final String visitStatus,
+        @RequestParam(value = "visitStatus", required = false) @ApiParam(name = "Status of visit. code from VIS_COMPLETE domain, e.g: Cancelled (CANC) or Scheduled (SCH)", example = "SCH") final String visitStatus,
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/prison/api/resource/BookingResource.java:882]: Confirmed by selecting all the different values in preprod, namely
```
select VISIT_STATUS, count(VISIT_STATUS) from OMS_OWNER.OFFENDER_VISITS group by VISIT_STATUS
```
which gives
```
CANC,2026931
EXP,110
HMPOP,7374
NORM,7078102
OFFEND,29646
SCH,5263401
VDE,1684
VISITOR,66619
```
The 110 `EXP` values look like they are dodgy data from 2007, all for `ALI` prison.
Also from NOMIS:
Screenshot from 2022-01-13 15-16-48
Note that `VIS_COMPLETE` values are a sub set of the `VIS_STS` values, but as a filter it makes more sense to only select on values that will be there.

File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/VisitInformation.java [hunk 4]
```diff
@@ -12,35 +12,40 @@
...
-        "LEFT JOIN REFERENCE_CODES RC5 ON RC5.DOMAIN = 'MOVE_CANC_RS' AND RC5.CODE = VISITOR.OUTCOME_REASON_CODE " +
-        "LEFT JOIN AGENCY_INTERNAL_LOCATIONS AIL ON VISIT.VISIT_INTERNAL_LOCATION_ID = AIL.INTERNAL_LOCATION_ID " +
-        "LEFT JOIN AGENCY_LOCATIONS AGY ON VISIT.AGY_LOC_ID = AGY.AGY_LOC_ID " +
-        "LEFT JOIN PERSONS P ON P.PERSON_ID = VISITOR.PERSON_ID " +
-        "ORDER BY VISIT.START_TIME DESC)")
+@Subselect("""
+    SELECT * FROM
+        (SELECT VISIT.OFFENDER_VISIT_ID VISIT_ID,
+                VISIT.VISIT_STATUS VISIT_STATUS,
+                RC1.DESCRIPTION VISIT_STATUS_DESCRIPTION,
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/VisitInformation.java:19]: add in the status description

File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/VisitInformation.java [hunk 5]
```diff
@@ -12,35 +12,40 @@
...
+                RC4.DESCRIPTION EVENT_OUTCOME_DESCRIPTION,
+                VISIT.START_TIME,
+                VISIT.END_TIME,
+                COALESCE(AIL.USER_DESC, AIL.DESCRIPTION, AGY.DESCRIPTION) LOCATION,
+                VISIT.VISIT_TYPE,
+                RC3.DESCRIPTION VISIT_TYPE_DESCRIPTION,
+                P.FIRST_NAME || ' ' || P.LAST_NAME LEAD_VISITOR
+           FROM OFFENDER_VISITS VISIT
+          INNER JOIN OFFENDER_BOOKINGS BOOKING ON BOOKING.OFFENDER_BOOK_ID = VISIT.OFFENDER_BOOK_ID AND BOOKING.ACTIVE_FLAG = 'Y'
+          INNER JOIN OFFENDER_VISIT_VISITORS VISIT_EXTRA ON VISIT.OFFENDER_VISIT_ID = VISIT_EXTRA.OFFENDER_VISIT_ID AND VISIT_EXTRA.OFFENDER_BOOK_ID IS NOT NULL
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/VisitInformation.java:36]: Join to visitors table on dummy row (`OFFENDER_BOOK_ID` not null) to get the extra visit information stored in that table instead.

File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/VisitInformation.java [hunk 6]
```diff
@@ -12,35 +12,40 @@
...
+           LEFT JOIN OFFENDER_VISIT_VISITORS LEAD_VISITOR ON VISIT.OFFENDER_VISIT_ID = LEAD_VISITOR.OFFENDER_VISIT_ID AND LEAD_VISITOR.GROUP_LEADER_FLAG = 'Y' AND LEAD_VISITOR.PERSON_ID IS NOT NULL
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/VisitInformation.java:37]: Join to visitors table for lead visitor

File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/VisitorInformation.java [hunk 7]
```diff
@@ -12,18 +12,23 @@
...
+@Subselect("""
+   SELECT * FROM
+       (SELECT P.PERSON_ID,
+               VISITOR.OFFENDER_VISIT_VISITOR_ID VISITOR_ID,
+               VISITOR.OFFENDER_VISIT_ID VISIT_ID,
+               P.FIRST_NAME,
+               P.LAST_NAME,
+               P.BIRTHDATE,
+               VISITOR.GROUP_LEADER_FLAG LEAD_VISITOR,
+               NVL(VISITOR.EVENT_OUTCOME,'ATT') EVENT_OUTCOME
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/VisitorInformation.java:24]: grab whether the visitor attended or not as well

File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/VisitorInformation.java [hunk 8]
```diff
@@ -12,18 +12,23 @@
...
+          FROM OFFENDER_VISIT_VISITORS VISITOR
+          JOIN PERSONS P ON P.PERSON_ID = VISITOR.PERSON_ID
+         WHERE VISITOR.OFFENDER_BOOK_ID IS NULL
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/VisitorInformation.java:27]: exclude the dummy row (`OFFENDER_BOOK_ID` is not null)

File: src/main/resources/db/migration/data/R__2_3__REFERENCE_CODES.sql [hunk 9]
```diff
@@ -8566,6 +8566,7 @@ INSERT INTO REFERENCE_CODES (DOMAIN, CODE, PARENT_DOMAIN, PARENT_CODE, DESCRIPTI
 INSERT INTO REFERENCE_CODES (DOMAIN, CODE, PARENT_DOMAIN, PARENT_CODE, DESCRIPTION, LIST_SEQ, ACTIVE_FLAG, SYSTEM_DATA_FLAG, EXPIRED_DATE) VALUES ('VIS_COMPLETE', 'OFFEND', null, null, 'Offender Completed Early', 99, 'Y', 'N', null);
 INSERT INTO REFERENCE_CODES (DOMAIN, CODE, PARENT_DOMAIN, PARENT_CODE, DESCRIPTION, LIST_SEQ, ACTIVE_FLAG, SYSTEM_DATA_FLAG, EXPIRED_DATE) VALUES ('VIS_COMPLETE', 'SCH', null, null, 'Scheduled', 99, 'Y', 'Y', null);
 INSERT INTO REFERENCE_CODES (DOMAIN, CODE, PARENT_DOMAIN, PARENT_CODE, DESCRIPTION, LIST_SEQ, ACTIVE_FLAG, SYSTEM_DATA_FLAG, EXPIRED_DATE) VALUES ('VIS_COMPLETE', 'VISITOR', null, null, 'Visitor Completed Early', 99, 'Y', 'N', null);
+INSERT INTO REFERENCE_CODES (DOMAIN, CODE, PARENT_DOMAIN, PARENT_CODE, DESCRIPTION, LIST_SEQ, ACTIVE_FLAG, SYSTEM_DATA_FLAG, EXPIRED_DATE) VALUES ('VIS_COMPLETE', 'VDE', null, null, 'Visitor Declined Entry', 99, 'Y', 'N', null);
```
Comment (petergphillips) [inline src/main/resources/db/migration/data/R__2_3__REFERENCE_CODES.sql:8569]: missing reference data that was added in 2015

File: src/test/java/uk/gov/justice/hmpps/prison/repository/jpa/repository/VisitorInformationRepositoryTest.java [hunk 10]
```diff
@@ -19,19 +19,19 @@
...
 
     @Autowired
     private VisitorRepository repository;
 
     @Test
     public void findAllByVisitId() {
-        var visits = repository.findAllByVisitId(-15L);
+        final var visits = repository.findAllByVisitId(-15L);
 
-        assertThat(visits).hasSize(3);
```
Comment (petergphillips) [inline src/test/java/uk/gov/justice/hmpps/prison/repository/jpa/repository/VisitorInformationRepositoryTest.java:31]: no point testing the size when we are testing all the values in the table anyway

File: src/main/resources/db/migration/data/R__5_9__OFFENDER_VISIT_VISITORS.sql [hunk 11]
```diff
@@ -1,11 +1,49 @@
...
+INSERT INTO OFFENDER_VISIT_VISITORS (OFFENDER_VISIT_VISITOR_ID, OFFENDER_VISIT_ID, OFFENDER_BOOK_ID, EVENT_ID, EVENT_STATUS, GROUP_LEADER_FLAG, EVENT_OUTCOME, OUTCOME_REASON_CODE, COMMENT_TEXT) VALUES (3135, -35, -6, 135, 'SCH', 'N', NULL, NULL, NULL);
+INSERT INTO OFFENDER_VISIT_VISITORS (OFFENDER_VISIT_VISITOR_ID, OFFENDER_VISIT_ID, OFFENDER_BOOK_ID, EVENT_ID, EVENT_STATUS, GROUP_LEADER_FLAG, EVENT_OUTCOME, OUTCOME_REASON_CODE, COMMENT_TEXT) VALUES (3136, -36, -6, 136, 'SCH', 'N', NULL, NULL, NULL);
+INSERT INTO OFFENDER_VISIT_VISITORS (OFFENDER_VISIT_VISITOR_ID, OFFENDER_VISIT_ID, OFFENDER_BOOK_ID, EVENT_ID, EVENT_STATUS, GROUP_LEADER_FLAG, EVENT_OUTCOME, OUTCOME_REASON_CODE, COMMENT_TEXT) VALUES (3137, -37, -6, 137, 'SCH', 'N', NULL, NULL, NULL);
 
+INSERT INTO OFFENDER_VISIT_VISITORS (OFFENDER_VISIT_ID,PERSON_ID,GROUP_LEADER_FLAG,OFFENDER_VISIT_VISITOR_ID,EVENT_OUTCOME,OFFENDER_BOOK_ID,EVENT_ID,CREATE_USER_ID,EVENT_STATUS) VALUES (-15,-1,'Y',-1,'ABS',null,400001,'ME','CANC');
```
Comment (petergphillips) [inline src/main/resources/db/migration/data/R__5_9__OFFENDER_VISIT_VISITORS.sql:38]: inserting the `EVENT_STATUS` here shouldn't be required, but still needed as other repository code still grabs the values from the lead visitor instead.  Will have to be tidied up at a later date...

**Review by andymarke — APPROVED** [review]
LGTM

**Pull request merged**